uniffi_bindgen/src/interface/record.rs

```diff
@@ -48,79 +48,107 @@ use anyhow::{bail, Result};
 
 use super::literal::{convert_default_value, Literal};
 use super::types::Type;
-use super::{APIConverter, ComponentInterface};
+use super::{APIConverter, ComponentInterface, CINode};
 
 /// Represents a "data class" style object, for passing around complex values.
 ///
 /// In the FFI these are represented as a byte buffer, which one side explicitly
 /// serializes the data into and the other serializes it out of. So I guess they're
 /// kind of like "pass by clone" values.
+#[derive(Debug)]
+pub struct Record<'a> {
+    pub(super) parent: &'a ComponentInterface,
+    pub(super) descr: &'a RecordDescr,
+}
+
 #[derive(Debug, Clone, Hash)]
-pub struct Record {
+pub struct RecordDescr {
     pub(super) name: String,
-    pub(super) fields: Vec<Field>,
+    pub(super) fields: Vec<FieldDescr>,
 }
 
-impl Record {
+impl<'a> Record<'a> {
     pub fn name(&self) -> &str {
-        &self.name
+        &self.descr.name
     }
 
-    pub fn fields(&self) -> Vec<&Field> {
-        self.fields.iter().collect()
+    pub fn fields(&self) -> Vec<Field<'_, Self>> {
+        self.descr.fields.iter().map(|f| Field { parent: self, descr: f }).collect()
     }
 
-    pub fn contains_object_references(&self, ci: &ComponentInterface) -> bool {
-        // *sigh* at the clone here, the relationship between a ComponentInterace
-        // and its contained types could use a bit of a cleanup.
-        ci.type_contains_object_references(&Type::Record(self.name.clone()))
+    pub fn contains_object_references(&self) -> bool {
+        self.fields().iter()
+            .any(|f| f.contains_object_references())
     }
 
-    pub fn contains_unsigned_types(&self, ci: &ComponentInterface) -> bool {
-        self.fields()
-            .iter()
-            .any(|f| ci.type_contains_unsigned_types(&f.type_))
+    pub fn contains_unsigned_types(&self, _ci: &ComponentInterface) -> bool {
+        self.descr.fields.iter()
+            .any(|f| self.ci().type_contains_unsigned_types(&f.type_))
     }
 }
 
-impl APIConverter<Record> for weedle::DictionaryDefinition<'_> {
-    fn convert(&self, ci: &mut ComponentInterface) -> Result<Record> {
+impl<'a> CINode for Record<'a> {
+    fn ci(&self) -> &ComponentInterface {
+        self.parent
+    }
+}
+
+impl APIConverter<RecordDescr> for weedle::DictionaryDefinition<'_> {
+    fn convert(&self, ci: &mut ComponentInterface) -> Result<RecordDescr> {
         if self.attributes.is_some() {
             bail!("dictionary attributes are not supported yet");
         }
         if self.inheritance.is_some() {
             bail!("dictionary inheritence is not supported");
         }
-        Ok(Record {
+        Ok(RecordDescr {
             name: self.identifier.0.to_string(),
             fields: self.members.body.convert(ci)?,
         })
     }
 }
 
 // Represents an individual field on a Record.
+#[derive(Debug)]
+pub struct Field<'a, Parent: CINode> {
```

Comment:
We re-use the `Field` struct in two places, for enum variants fields and record fields, hence the generic.

Which makes me wonder...I wonder if enum variants could actually contain a wrapped `Record` in the same way that errors contain a wrapped `Enum`...